lhotse/workflows/meeting_simulation/base.py

```diff
@@ -83,6 +89,120 @@ def reverberate(self, cuts: CutSet, *rirs: RecordingSet) -> CutSet:
+            f"Got: {num_speakers_per_meeting}"
+        )
+        assert all(p > 0.0 for p in speaker_count_probs), (
+            "The probabilities of the number of speakers per meeting must be greater than 0. "
+            f"Got: {speaker_count_probs}"
+        )
+        assert sum(speaker_count_probs) == 1.0, (
+            "The probabilities of the number of speakers per meeting must sum to 1. "
+            f"Got: {speaker_count_probs}"
+        )
+        assert len(num_speakers_per_meeting) == len(
+            speaker_count_probs
+        ), "The number of speakers per meeting and the number of probabilities must be the same."
+
+        # Create samplers for each bucket. We create this as a dict so that we can
+        # efficiently remove items and also randomly sample items in constant time.
+        # It also supports the len() function in constant time.
+        # Note that a Python list is not a good choice here, because removing items
+        # from a list is O(n). A set is also not a good choice, because randomly
+        # sampling items from a set is O(n).
+        self.samplers = {}
+        for spk, spk_cuts in tqdm(
+            groupby(cuts, lambda cut: cut.supervisions[0].speaker),
+            desc="Creating samplers for each speaker...",
+        ):
+            sampler = DynamicCutSampler(
+                CutSet.from_cuts(list(spk_cuts)).repeat(
+                    times=num_repeats, preserve_id=False
+                ),
+                max_duration=max_duration_per_speaker,
+                max_cuts=max_utterances_per_speaker,
+                shuffle=True,
+                seed=seed,
+            )
+            self.samplers[spk] = sampler
+
+        self.num_speakers_per_meeting = num_speakers_per_meeting
+        self.speaker_count_probs = speaker_count_probs
+
+        self.npr = np.random.RandomState(seed)
+        self.rng = random.Random(seed)
+        self._remaining_meetings = num_meetings
+
+    def __iter__(self):
+        for sampler in self.samplers.values():
+            iter(sampler)
+        return self
+
+    def __next__(self):
+        # If we have sampled enough meetings, stop.
+        if self._remaining_meetings is not None and self._remaining_meetings == 0:
+            raise StopIteration()
+
+        # If we don't have enough speakers, stop.
+        if len(self.samplers) < min(self.num_speakers_per_meeting):
+            raise StopIteration()
+
+        # Sample the number of speakers for this meeting.
+        N = min(
+            self.npr.choice(self.num_speakers_per_meeting, p=self.speaker_count_probs),
+            len(self.samplers),
+        )
+
+        # Sample speakers.
+        this_batch_spk_ids = self.rng.sample(list(self.samplers.keys()), N)
+        utterances = CutSet.from_cuts([])
+        for spk_id in this_batch_spk_ids:
+            sampler = self.samplers[spk_id]
+            try:
+                this_batch = next(sampler)
```

Comment:
@pzelasko An issue I am facing here is that the whole sampler gets exhausted after sampling just 1 batch. I'm not sure why this is happening. Could you take a look?

Reply:
Okay very basic mistake --- forgot to sort the cuts by speaker id before groupby!